ibm/resource_ibm_cloudant_database.go

```diff
@@ -0,0 +1,189 @@
+	if _, ok := d.GetOk("partitioned"); ok {
+		putDatabaseOptions.SetPartitioned(d.Get("partitioned").(bool))
+	}
+	if _, ok := d.GetOk("q"); ok {
+		putDatabaseOptions.SetQ(int64(d.Get("q").(int)))
+	}
+
+	_, response, err := cloudantClient.PutDatabaseWithContext(context, putDatabaseOptions)
+	if err != nil {
+		log.Printf("[DEBUG] PutDatabaseWithContext failed %s\n%s", err, response)
+		return diag.FromErr(fmt.Errorf("PutDatabaseWithContext failed %s\n%s", err, response))
+	}
+
+	d.SetId(fmt.Sprintf("%s/%s", instanceCRN, dbName))
+
+	return resourceIbmCloudantDatabaseRead(context, d, meta)
+}
+
+func resourceIbmCloudantDatabaseRead(context context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	parts, err := idParts(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	instanceCRN, dbName := strings.Join(parts[:len(parts)-1], "/"), parts[len(parts)-1]
+	cUrl, err := getCloudantInstanceUrl(instanceCRN, meta)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	cloudantClient, err := getCloudantClientForUrl(cUrl, meta)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	getDatabaseInformationOptions := cloudantClient.NewGetDatabaseInformationOptions(dbName)
+
+	databaseInformation, response, err := cloudantClient.GetDatabaseInformationWithContext(context, getDatabaseInformationOptions)
+	if err != nil {
+		if response != nil && response.StatusCode == 404 {
+			d.SetId("")
+			return nil
+		}
+		log.Printf("[DEBUG] GetDatabaseInformationWithContext failed %s\n%s", err, response)
+		return diag.FromErr(fmt.Errorf("GetDatabaseInformationWithContext failed %s\n%s", err, response))
+	}
+
+	d.Set("instance_crn", instanceCRN)
+
+	if err = d.Set("db", *databaseInformation.DbName); err != nil {
+		return diag.FromErr(fmt.Errorf("Error setting db: %s", err))
+	}
+
+	if _, ok := d.GetOk("partitioned"); ok {
+		d.Set("partitioned", true)
+	}
+
+	if qData, ok := d.GetOk("q"); ok {
+		d.Set("q", qData.(int))
+	}
+
+	return nil
+}
+
+func resourceIbmCloudantDatabaseDelete(context context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	parts, err := idParts(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	instanceCRN, dbName := strings.Join(parts[:len(parts)-1], "/"), parts[len(parts)-1]
+	cUrl, err := getCloudantInstanceUrl(instanceCRN, meta)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	cloudantClient, err := getCloudantClientForUrl(cUrl, meta)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	deleteDatabaseOptions := cloudantClient.NewDeleteDatabaseOptions(dbName)
+
+	_, response, err := cloudantClient.DeleteDatabaseWithContext(context, deleteDatabaseOptions)
+	if err != nil {
+		log.Printf("[DEBUG] DeleteDatabaseWithContext failed %s\n%s", err, response)
+		return diag.FromErr(fmt.Errorf("DeleteDatabaseWithContext failed %s\n%s", err, response))
+	}
+
+	d.SetId("")
+
+	return nil
+}
+
+func getCloudantInstanceUrl(instanceCRN string, meta interface{}) (string, error) {
```

Comment:
bit of a shame to have to duplicate a lot of this functionality from the cloudant isntance resource, but I can see that we start from a different place and I can live with it.

Reply:
Yes, this is pretty hacky, especially part where we calling "read" after "create" and this approach doesn't reuse client at all and just throws away old one and creates new one. The problem is that we are not using globally shared client for obvious reasons and there is no place to reliably store Cloudant instance's client (if instance is a part of terraform config at all, it might be managed externally) and same time guarantee correct sequential order of resource calling 😞 

Pretty much the only alternative I can think of is to switch to global client similar to the rest of the resources in this provider, but use mutex to manage url setter. Downside ofc is that we can accidentally lock the whole system with unexpected response from any of controllers, so I'm hesitant to go there.